File: 3.12/main.c
#include "../nob.h"

ll Part1(const char *line) {
    int len = strlen(line) - 1;
    ll max_joltage = 0;

    for (int i = 0; i < len; i++) {
        for (int j = i + 1; j < len; j++) {
            ll joltage = (line[i] - '0') * 10 + (line[j] - '0');
            if (joltage > max_joltage) max_joltage = joltage;
        }
    }
    
    return max_joltage;
}

ll Part2(const char *line) {
    int len = strlen(line) - 1;
    if (len < 12) return 0;

    int digits_to_remove = len - 12;

    char stack[len+1];
    int top = 0;

    for (int i = 0; i < len; i++) {
    
        char current = line[i];

   		while (top > 0 && digits_to_remove > 0 && stack[top - 1] < current) {
            top--;
            digits_to_remove--;
        }

        stack[top++] = current;
    }

    top -= digits_to_remove;

    ll joltage = 0;

    for (int i = 0; i < 12; i++) {
        joltage = joltage * 10 + (stack[i] - '0');
    }

    return joltage;
}

int main(int argc, char *argv[]) {
    
    FILE *file = open_input_or_die(argc, argv);
    
    char *line = NULL;
    size_t len = 0;
    ssize_t read;

    ll part1 = 0;
    ll part2 = 0;

    while ((read = getline_or_die(&line, &len, file)) != -1) {
        if (strlen(line) > 0) {
            part1 += Part1(line);
            part2 += Part2(line);
        }
    }

    printf("Part 1: %lld\n", part1);
    printf("Part 2: %lld\n", part2);
    	
    free(line);
    fclose(file);
	
    return 0;
}

File: 3.12/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader};

fn part1(line: &str) -> i64 {
    let len = line.len(); // FIXED
    let bytes = line.as_bytes();
    let mut max_joltage: i64 = 0;

    for i in 0..len {
        for j in (i + 1)..len {
            let joltage =
                ((bytes[i] - b'0') as i64) * 10 + (bytes[j] - b'0') as i64;
            if joltage > max_joltage {
                max_joltage = joltage;
            }
        }
    }

    max_joltage
}

fn part2(line: &str) -> i64 {
    let len = line.len(); // FIXED
    if len < 12 {
        return 0;
    }

    let mut digits_to_remove = len - 12;
    let mut stack: Vec<u8> = Vec::with_capacity(len);

    for &current in line.as_bytes() {
        while !stack.is_empty()
            && digits_to_remove > 0
            && *stack.last().unwrap() < current
        {
            stack.pop();
            digits_to_remove -= 1;
        }
        stack.push(current);
    }

    let final_len = stack.len() - digits_to_remove;
    stack.truncate(final_len);

    let mut joltage: i64 = 0;
    for i in 0..12 {
        joltage = joltage * 10 + (stack[i] - b'0') as i64;
    }

    joltage
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();

    if args.len() != 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    let file = File::open(&args[1])?;
    let reader = BufReader::new(file);

    let mut part1_sum: i64 = 0;
    let mut part2_sum: i64 = 0;

    for line in reader.lines() {
        let line = line?;
        if !line.is_empty() {
            part1_sum += part1(&line);
            part2_sum += part2(&line);
        }
    }

    println!("Part 1: {}", part1_sum);
    println!("Part 2: {}", part2_sum);

    Ok(())
}
